File: AnarAndroid/app/src/main/java/app/anar/android/core/SingboxConfig.kt
package app.anar.android.core

import org.json.JSONArray
import org.json.JSONObject

/**
 * Generates a sing-box (v1.13) JSON config for libbox. Mirrors the macOS
 * generator, whose output was validated against a real sing-box build.
 */
object SingboxConfig {

    const val PROXY_TAG = "proxy"

    fun generate(profile: ProxyProfile, settings: AppSettings): String {
        val root = JSONObject()

        root.put("log", JSONObject().put("level", settings.logLevel).put("timestamp", true))

        // remote = through proxy, local = OS resolver (direct) for bootstrapping.
        val dnsServers = JSONArray()
            .put(dnsServer(settings.dnsServer))
            .put(JSONObject().put("type", "local").put("tag", "local"))
        root.put("dns", JSONObject().put("servers", dnsServers).put("strategy", "prefer_ipv4"))

        // TUN inbound — libbox provides the fd via the VpnService platform interface.
        val tun = JSONObject()
            .put("type", "tun")
            .put("tag", "tun-in")
            .put("address", JSONArray().put("172.18.0.1/30"))
            .put("auto_route", true)
            .put("strict_route", false)
            .put("stack", "gvisor")
        root.put("inbounds", JSONArray().put(tun))

        if (profile.type == ProxyType.WIREGUARD) {
            root.put("endpoints", JSONArray().put(wireguardEndpoint(profile)))
            root.put("outbounds", JSONArray().put(direct()))
        } else {
            root.put("outbounds", JSONArray().put(outbound(profile)).put(direct()))
        }

        val rules = JSONArray()
            .put(JSONObject().put("action", "sniff"))
            .put(JSONObject().put("protocol", "dns").put("action", "hijack-dns"))
        if (settings.bypassPrivate) {
            rules.put(JSONObject().put("ip_is_private", true).put("outbound", "direct"))
        }
        root.put("route", JSONObject()
            .put("rules", rules)
            .put("final", PROXY_TAG)
            .put("auto_detect_interface", true)
            .put("default_domain_resolver", "local"))

        root.put("experimental", JSONObject().put("clash_api",
            JSONObject().put("external_controller", "127.0.0.1:${settings.clashApiPort}")))

        return root.toString(2)
    }

    private fun direct() = JSONObject().put("type", "direct").put("tag", "direct")

    private fun dnsServer(spec: String): JSONObject {
        var type = "udp"
        var server = spec
        val idx = spec.indexOf("://")
        if (idx >= 0) {
            type = spec.substring(0, idx).lowercase()
            server = spec.substring(idx + 3)
        }
        if (type != "https") server.indexOf('/').takeIf { it >= 0 }?.let { server = server.substring(0, it) }
        return when (type) {
            "https", "h3" -> {
                val host = server.substringBefore('/')
                val path = if (server.contains('/')) "/" + server.substringAfter('/') else "/dns-query"
                JSONObject().put("type", type).put("tag", "remote").put("server", host)
                    .put("path", path).put("domain_resolver", "local")
            }
            "quic" -> JSONObject().put("type", "quic").put("tag", "remote").put("server", server)
            "tls" -> JSONObject().put("type", "tls").put("tag", "remote").put("server", server)
            else -> JSONObject().put("type", "udp").put("tag", "remote").put("server", server)
        }
    }

    private fun outbound(p: ProxyProfile): JSONObject {
        val o = JSONObject()
            .put("tag", PROXY_TAG)
            .put("server", p.server)
            .put("server_port", p.port)
        when (p.type) {
            ProxyType.VMESS -> {
                o.put("type", "vmess").put("uuid", p.uuid)
                    .put("security", p.encryption.ifEmpty { "auto" }).put("alter_id", p.alterId)
                mergeTransport(o, p); mergeTLS(o, p)
            }
            ProxyType.VLESS -> {
                o.put("type", "vless").put("uuid", p.uuid)
                if (p.flow.isNotEmpty()) o.put("flow", p.flow)
                mergeTransport(o, p); mergeTLS(o, p)
            }
            ProxyType.TROJAN -> {
                o.put("type", "trojan").put("password", p.password)
                mergeTransport(o, p); mergeTLS(o, p)
            }
            ProxyType.SHADOWSOCKS ->
                o.put("type", "shadowsocks").put("method", p.method).put("password", p.password)
            ProxyType.HYSTERIA2 -> {
                o.put("type", "hysteria2").put("password", p.password)
                if (p.obfs.isNotEmpty()) {
                    o.put("obfs", JSONObject().put("type", p.obfs).put("password", p.obfsPassword))
                }
                o.put("tls", tlsBlock(p))
            }
            ProxyType.TUIC -> {
                o.put("type", "tuic").put("uuid", p.uuid).put("password", p.password)
                    .put("congestion_control", p.congestionControl.ifEmpty { "bbr" })
                    .put("udp_relay_mode", p.udpRelayMode.ifEmpty { "native" })
                o.put("tls", tlsBlock(p))
            }
            ProxyType.SOCKS -> o.put("type", "socks")
            ProxyType.HTTP -> o.put("type", "http")
            ProxyType.WIREGUARD -> {} // handled via endpoint
        }
        return o
    }

    private fun wireguardEndpoint(p: ProxyProfile): JSONObject {
        val peer = JSONObject()
            .put("address", p.server)
            .put("port", p.port)
            .put("public_key", p.peerPublicKey)
            .put("allowed_ips", JSONArray().put("0.0.0.0/0").put("::/0"))
        if (p.preSharedKey.isNotEmpty()) peer.put("pre_shared_key", p.preSharedKey)
        if (p.reserved.isNotEmpty()) peer.put("reserved", JSONArray(p.reserved))
        val addresses = if (p.localAddresses.isEmpty()) listOf("172.16.0.2/32") else p.localAddresses
        return JSONObject()
            .put("type", "wireguard")
            .put("tag", PROXY_TAG)
            .put("address", JSONArray(addresses))
            .put("private_key", p.privateKey)
            .put("peers", JSONArray().put(peer))
    }

    private fun mergeTransport(o: JSONObject, p: ProxyProfile) {
        when (p.network) {
            "ws" -> {
                val t = JSONObject().put("type", "ws")
                if (p.path.isNotEmpty()) t.put("path", p.path)
                if (p.hostHeader.isNotEmpty()) t.put("headers", JSONObject().put("Host", p.hostHeader))
                o.put("transport", t)
            }
            "grpc" -> o.put("transport", JSONObject().put("type", "grpc").put("service_name", p.path))
            "http", "h2" -> {
                val t = JSONObject().put("type", "http")
                if (p.hostHeader.isNotEmpty()) t.put("host", JSONArray().put(p.hostHeader))
                if (p.path.isNotEmpty()) t.put("path", p.path)
                o.put("transport", t)
            }
        }
    }

    private fun mergeTLS(o: JSONObject, p: ProxyProfile) {
        if (p.tls || p.reality) o.put("tls", tlsBlock(p))
    }

    private fun tlsBlock(p: ProxyProfile): JSONObject {
        val tls = JSONObject().put("enabled", true)
        tls.put("server_name", p.sni.ifEmpty { p.server })
        if (p.insecure) tls.put("insecure", true)
        if (p.alpn.isNotEmpty()) tls.put("alpn", JSONArray(p.alpn))
        if (p.fingerprint.isNotEmpty()) {
            tls.put("utls", JSONObject().put("enabled", true).put("fingerprint", p.fingerprint))
        }
        if (p.reality) {
            tls.put("reality", JSONObject().put("enabled", true)
                .put("public_key", p.publicKey).put("short_id", p.shortId))
            if (p.fingerprint.isEmpty()) {
                tls.put("utls", JSONObject().put("enabled", true).put("fingerprint", "chrome"))
            }
        }
        return tls
    }
}

File: Anar/Anar/Core/SingboxConfig.swift
import Foundation

/// Builds a sing-box (v1.13) JSON configuration for a selected profile.
enum SingboxConfig {

    static let proxyTag = "proxy"

    static func generate(profile: ProxyProfile, settings: AppSettings) throws -> Data {
        var root: [String: Any] = [:]

        root["log"] = ["level": settings.logLevel, "timestamp": true]

        // "remote" tunnels DNS through the proxy; "local" uses the OS resolver
        // (direct) and bootstraps the proxy server's own domain — this avoids the
        // resolve-loop without a forbidden detour-to-direct.
        root["dns"] = [
            "servers": [
                dnsServer(settings.dnsServer),
                ["type": "local", "tag": "local"],
            ],
            "strategy": "prefer_ipv4",
        ]

        // Inbound
        switch settings.mode {
        case .tun:
            root["inbounds"] = [[
                "type": "tun",
                "tag": "tun-in",
                "address": ["172.18.0.1/30"],
                "auto_route": true,
                "strict_route": true,
                "stack": "gvisor",
            ]]
        case .proxy:
            root["inbounds"] = [[
                "type": "mixed",
                "tag": "mixed-in",
                "listen": "127.0.0.1",
                "listen_port": settings.mixedPort,
            ]]
        }

        // Outbounds / endpoints
        if profile.type == .wireguard {
            root["endpoints"] = [wireguardEndpoint(profile)]
            root["outbounds"] = [["type": "direct", "tag": "direct"]]
        } else {
            root["outbounds"] = [try outbound(profile), ["type": "direct", "tag": "direct"]]
        }

        // Route
        var rules: [[String: Any]] = [
            ["action": "sniff"],
            ["protocol": "dns", "action": "hijack-dns"],
        ]
        var ruleSets: [[String: Any]] = []

        if settings.routingRule != .global {
            rules.append(["ip_is_private": true, "outbound": "direct"])
        }
        if settings.adBlock {
            ruleSets.append(remoteRuleSet("geosite-ads", "geosite-category-ads-all"))
            rules.append(["rule_set": ["geosite-ads"], "action": "reject"])
        }
        if settings.routingRule == .bypassIran {
            ruleSets.append(remoteRuleSet("geoip-ir", "geoip-ir"))
            ruleSets.append(remoteRuleSet("geosite-ir", "geosite-ir"))
            rules.append(["rule_set": ["geoip-ir", "geosite-ir"], "outbound": "direct"])
        }

        var route: [String: Any] = [
            "rules": rules,
            "final": proxyTag,
            "auto_detect_interface": true,
            "default_domain_resolver": "local",
        ]
        if !ruleSets.isEmpty { route["rule_set"] = ruleSets }
        root["route"] = route

        // Live stats API
        root["experimental"] = [
            "clash_api": [
                "external_controller": "127.0.0.1:\(settings.clashApiPort)",
            ],
        ]

        return try JSONSerialization.data(withJSONObject: root, options: [.prettyPrinted, .sortedKeys])
    }

    /// A remote .srs rule-set, downloaded through the proxy (reliable in Iran).
    private static func remoteRuleSet(_ tag: String, _ name: String) -> [String: Any] {
        [
            "type": "remote",
            "tag": tag,
            "format": "binary",
            "url": "https://raw.githubusercontent.com/Chocolate4U/Iran-sing-box-rules/rule-set/\(name).srs",
            "download_detour": proxyTag,
        ]
    }

    /// Builds a sing-box 1.12+ DNS server object from a "scheme://host" string.
    /// Examples: "tls://8.8.8.8", "https://dns.google/dns-query", "1.1.1.1".
    private static func dnsServer(_ spec: String) -> [String: Any] {
        var type = "udp"
        var server = spec
        if let r = spec.range(of: "://") {
            type = String(spec[..<r.lowerBound]).lowercased()
            server = String(spec[r.upperBound...])
        }
        // Keep only the host for tls/udp/quic; https keeps the full host.
        if type != "https", let slash = server.firstIndex(of: "/") {
            server = String(server[..<slash])
        }
        switch type {
        case "https": return ["type": "https", "tag": "remote", "server": serverHost(server), "path": serverPath(server), "domain_resolver": "local"]
        case "tls": return ["type": "tls", "tag": "remote", "server": server]
        case "quic": return ["type": "quic", "tag": "remote", "server": server]
        case "h3": return ["type": "h3", "tag": "remote", "server": serverHost(server), "path": serverPath(server), "domain_resolver": "local"]
        default: return ["type": "udp", "tag": "remote", "server": server]
        }
    }

    private static func serverHost(_ s: String) -> String {
        if let slash = s.firstIndex(of: "/") { return String(s[..<slash]) }
        return s
    }
    private static func serverPath(_ s: String) -> String {
        if let slash = s.firstIndex(of: "/") { return String(s[slash...]) }
        return "/dns-query"
    }

    // MARK: - Outbound builders

    private static func outbound(_ p: ProxyProfile) throws -> [String: Any] {
        var o: [String: Any] = [
            "tag": proxyTag,
            "server": p.server,
            "server_port": p.port,
        ]
        switch p.type {
        case .vmess:
            o["type"] = "vmess"
            o["uuid"] = p.uuid
            o["security"] = p.encryption.isEmpty ? "auto" : p.encryption
            o["alter_id"] = p.alterId
            mergeTransport(&o, p); mergeTLS(&o, p)
        case .vless:
            o["type"] = "vless"
            o["uuid"] = p.uuid
            if !p.flow.isEmpty { o["flow"] = p.flow }
            mergeTransport(&o, p); mergeTLS(&o, p)
        case .trojan:
            o["type"] = "trojan"
            o["password"] = p.password
            mergeTransport(&o, p); mergeTLS(&o, p)
        case .shadowsocks:
            o["type"] = "shadowsocks"
            o["method"] = p.method
            o["password"] = p.password
        case .hysteria2:
            o["type"] = "hysteria2"
            o["password"] = p.password
            if !p.obfs.isEmpty {
                o["obfs"] = ["type": p.obfs, "password": p.obfsPassword]
            }
            o["tls"] = tlsBlock(p, forceEnabled: true)
        case .tuic:
            o["type"] = "tuic"
            o["uuid"] = p.uuid
            o["password"] = p.password
            o["congestion_control"] = p.congestionControl.isEmpty ? "bbr" : p.congestionControl
            o["udp_relay_mode"] = p.udpRelayMode.isEmpty ? "native" : p.udpRelayMode
            o["tls"] = tlsBlock(p, forceEnabled: true)
        case .socks:
            o["type"] = "socks"
        case .http:
            o["type"] = "http"
        case .wireguard:
            throw SingboxError.handledAsEndpoint
        }
        return o
    }

    private static func wireguardEndpoint(_ p: ProxyProfile) -> [String: Any] {
        var peer: [String: Any] = [
            "address": p.server,
            "port": p.port,
            "public_key": p.peerPublicKey,
            "allowed_ips": ["0.0.0.0/0", "::/0"],
            "persistent_keepalive_interval": 25,
        ]
        if !p.preSharedKey.isEmpty { peer["pre_shared_key"] = p.preSharedKey }
        if !p.reserved.isEmpty { peer["reserved"] = p.reserved }
        return [
            "type": "wireguard",
            "tag": proxyTag,
            "mtu": 1408,
            "address": p.localAddresses.isEmpty ? ["172.16.0.2/32"] : p.localAddresses,
            "private_key": p.privateKey,
            "peers": [peer],
        ]
    }

    // MARK: - Shared blocks

    private static func mergeTransport(_ o: inout [String: Any], _ p: ProxyProfile) {
        switch p.network {
        case "ws":
            var t: [String: Any] = ["type": "ws"]
            if !p.path.isEmpty { t["path"] = p.path }
            if !p.hostHeader.isEmpty { t["headers"] = ["Host": p.hostHeader] }
            o["transport"] = t
        case "grpc":
            o["transport"] = ["type": "grpc", "service_name": p.path]
        case "http", "h2":
            var t: [String: Any] = ["type": "http"]
            if !p.hostHeader.isEmpty { t["host"] = [p.hostHeader] }
            if !p.path.isEmpty { t["path"] = p.path }
            o["transport"] = t
        default:
            break // tcp: no transport block
        }
    }

    private static func mergeTLS(_ o: inout [String: Any], _ p: ProxyProfile) {
        guard p.tls || p.reality else { return }
        o["tls"] = tlsBlock(p, forceEnabled: true)
    }

    private static func tlsBlock(_ p: ProxyProfile, forceEnabled: Bool) -> [String: Any] {
        var tls: [String: Any] = ["enabled": true]
        let serverName = p.sni.isEmpty ? p.server : p.sni
        tls["server_name"] = serverName
        if p.insecure { tls["insecure"] = true }
        if !p.alpn.isEmpty { tls["alpn"] = p.alpn }
        if !p.fingerprint.isEmpty {
            tls["utls"] = ["enabled": true, "fingerprint": p.fingerprint]
        }
        if p.reality {
            tls["reality"] = [
                "enabled": true,
                "public_key": p.publicKey,
                "short_id": p.shortId,
            ]
            // REALITY requires uTLS; default to chrome if none specified.
            if p.fingerprint.isEmpty {
                tls["utls"] = ["enabled": true, "fingerprint": "chrome"]
            }
        }
        return tls
    }

    enum SingboxError: Error { case handledAsEndpoint }
}
